src/filter/aech0r/aech0r.cpp

```diff
@@ -0,0 +1,306 @@
+ * to the GNU Public License for more details.
+ *
+ * You should  have received  a copy of  the GNU Public  License along
+ * with this source code; if  not, write to: Free Software Foundation,
+ * Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+ *
+ */
+#include "frei0r.hpp"
+#include "frei0r_math.h"
+
+#include <string.h>
+#include <climits>
+
+////// Uncomment to force non optimisation version
+//~ #undef __SSE2__
+
+////// TODO / IDEAS / ... //////
+// IDEA - directionnal echo ? (X/Y like parameter )
+// TODO RGB gradiant by fading influence need more love (See '//Fade by color layers')!
+// FIXME SSE2 version doesnt support RGB fading influence !
+
+// FIXME (Veejay specifics?) on activate/desactivate/activate/..., some buffers must be cleared!
+
+// EXPORE ME -
+//~ if((skip_count++)>m_skip) {
+//~ -      skip_count = 0;
+//~ -      m_factor += (factor * 64);
+//~ [...]
+//~ }
+
+// EXPLORE ME a very high m_factor value give some interresting color result (      m_factor += (factor * 64) * m_skip;)
+
+
+/* Intrinsic declarations */
+#if defined(__SSE2__)
+#include <emmintrin.h>
+// TODO mmx support and others
+//  #elif defined(__MMX__)
+// #include <mmintrin.h>
+#endif
+
+#define SKIP_MAX_IMAGES 8
+#define M_FACTOR_MAV    127
+
+union px_t {
+  uint32_t u;
+  unsigned char c[4]; // 0=B, 1=G,2=R,3=A ? i think :P
+};
+
+class aech0r : public frei0r::filter {
+private:
+  f0r_param_double factor;
+  //~ f0r_param_double factor_r;  //Fade by color layers
+  //~ f0r_param_double factor_g;
+  //~ f0r_param_double factor_b;
+  f0r_param_double strobe_period;
+  bool bright;
+  bool flag_r;
+  bool flag_g;
+  bool flag_b;
+
+
+  //~ f0r_param_double fade_rgb;  //Fade by color layers
+  //~ f0r_param_double flag_rgb;
+
+  unsigned int m_factor;
+
+  //~ unsigned int m_factor_r;  //Fade by color layers
+  //~ unsigned int m_factor_g;
+  //~ unsigned int m_factor_b;
+
+  //~ unsigned int m_flag_rgb;  //Fade by color layers
+  //~ unsigned int m_flag_r;
+  //~ unsigned int m_flag_g;
+  //~ unsigned int m_flag_b;
+
+  unsigned int m_skip;
+  unsigned int m_skip_count;
+
+  bool firsttime;
+
+  //~ unsigned int m_rgb; //Fade by color layers
+
+#ifdef __SSE2__
+  long long int m_factor_sse2;
+
+  inline void tracesse2_add(uint32_t* out, const uint32_t* in);
+  inline void tracesse2_sub(uint32_t* out, const uint32_t* in);
+#else
+  unsigned int m_factor_r;
+  unsigned int m_factor_g;
+  unsigned int m_factor_b;
+
+  inline void trace_add(uint32_t* out, const uint32_t* in);
+  inline void trace_sub(uint32_t* out, const uint32_t* in);
+#endif
+public:
+
+  aech0r(unsigned int width, unsigned int height) {
+
+    factor = 13;
+    strobe_period = 0;
+    bright = false;
+    
+    flag_r = false;
+    flag_g = false;
+    flag_b = false;
+    //~ fade_rgb = ?  //Fade by color layers
+
+    firsttime = true;
+    m_skip_count = 0;
+
+    register_param(factor, "Fade Factor", "Disappearance rate of the echo: from 0 to 127"); // 0 No fade, 1 No Trace
```

Comment:
All parameters must operate in and only in the range [0, 1].

Reply:
>> All parameters must operate in and only in the range [0, 1].

Internally speacking, for sure.
But, when the parameter act has a selector (enumeration), describing the range is necessary. 
 (ex ndvi filter : https://github.com/dyne/frei0r/blob/8639a58cea8361f40dfb7f8d502b3b2b3261e8b5/src/filter/ndvi/ndvi.cpp#L95-L96)

Elsewhere, like aech0r / factor I agree; knowing this range doesn't help to better manipulate the effect.

Reply:
Sorry I should have specified numeric parameter. The example you refer to is a string. Your change to the parameter description alone is not enough because the default for factor is 7.  Perhaps when using the factor member variable you want to multiple it by 127 and change the default to 7.0/127.0 ?

Reply:
7d833e9  >  filter/aechor : Fix factor default value